3.12/README.md

```diff
@@ -0,0 +1,146 @@
+# What will 3.12 look like?
+
+Author: Mark Shannon
+
+## Fundamentals
+
+* We can't speedup opaque blobs of C code.
+    * And we don't want use some elaborate mechanism like PyPy or GraalPython to understand the C code.
+* We can speedup code that runs in the interpreter
+* We can reduce the overhead of memory management, by doing less of it, and by doing it more efficiently.
+* The fastest way to do something is to not do it all.
+
+### This means that we want:
+
+* C code to consist of builtin functions that are well typed and don't call back in Python (much).
+* Everything else in bytecode (99.9% Python, but we might need a few key trampolines in hand written bytecode).
+* To be able to execute Python code faster.
+* To remove incidental overhead due to poor memory layout, reference counting and cycle GC.
+
+## Executing Python faster
+
+This is the most important part of a fast VM, and the most complex (unless we want a fully concurrent, parallel cycle GC).
+Ultimately we will need a JIT compiler, but it is clear from the relatively poor performance of Jython and IronPython that simply compiling to machine code is not magic pixie dust.
+Remembering that not doing something at all is always preferable to doing it faster, we want to optimize by removing overheads, including type-checks, boxing and unboxing, reference counting operations, object allocation.
+
+### Region selection
+
+The simplest region to optimize is a linear trace; all instructions dominate the latter ones, and there are no merge points.
+Traces can be stitched together to form larger regions, or even recompiled as a larger region (see HHVM).
+
+Pyston and Cinder choose single Python functions as their base region of optimization. Even with inlining, this is quite limiting.
+
+My PhD thesis compares a tracing interpreter (HotPy) with a method-at-a-time JIT compiler that did little or no specialization (Unladen Swallow); the interpreter wins easily.
```

Comment:
This seems like not a very meaningful comparison, or at least it's not really clear here what conclusions you think ought to be drawn from it. "JIT compiler that did little or no specialization" seems like setting up a straw man.

Reply:
That's what Unladen Swallow was. No straw man.
The point (which I will clarify) is that specialization is more important than compiling to machine code.

Reply:
The clarification will be useful, and is undoubtedly true. The location of the comment makes it seem like it's intended be a commentary on region selection :)

Reply:
Actually now I look, this doesn't belong here.